src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs

```diff
@@ -26,11 +20,48 @@ internal BoundExpression WithSuppression(bool suppress = true)
             // There is no scenario where suppression goes away
             Debug.Assert(suppress || !this.IsSuppressed);
 
-            var result = ShallowClone();
+            var result = (BoundExpression)MemberwiseClone();
             result.IsSuppressed = suppress;
             return result;
         }
 
+        internal BoundExpression WithWasConverted()
+        {
+#if DEBUG
+            // We track the WasConverted flag for locals and parameters only, as many other
```

Comment:
> We track the WasConverted flag for locals and parameters only [](start = 14, length = 62)

I didn't understand this. Could you elaborate? What is an example of special behavior that prevents such tracking from working?
Also, in `BindToNaturalType`, we're using `WithWasConverted` on converted switch and tuple expressions. That seems useful even though they are not locals or parameters.

Reply:
We have special nodes for out parameters, deconstruction variables, and a bunch of others.  Also, the code for doing this causes an extra duplication of every node that is tracked that way, and it would double the space allocated for bound nodes if we did it for everything (in DEBUG).

`WithWasConverted` is a no-op on anything except locals and parameters.

---
In reply to: [301793834](https://github.com/dotnet/roslyn/pull/37052#discussion_r301793834) [](ancestors = 301793834)

Reply:
Also lambdas, method groups.  Also `null` does not require conversion to a target type in some circumstances, such as `null==null`.

---
In reply to: [302313036](https://github.com/dotnet/roslyn/pull/37052#discussion_r302313036) [](ancestors = 302313036,301793834)